File: haskell/HumanPlayer.hs
module HumanPlayer where
import System.IO

import Games

humanPlayer :: Game b m => Player b m
humanPlayer game =
  do --putStrLn $ "Valid moves: " ++ show (moves (turn game) (state game))
     putStr $ "Move for player " ++ show (turn game + 1) ++ ": "
     hFlush stdout
     moveTxt <- getLine
     case readMove (turn game) moveTxt (board game) of
       Left move -> 
         if elem move $ moves (turn game) (board game)
         then return move
         else do putStrLn $ "Invalid move"-- ++ show move
                 humanPlayer game
       Right msg ->
         do putStrLn msg
            humanPlayer game
     

File: cpp/src/HumanPlayer.cpp

#include "HumanPlayer.hpp"

#include <string.h>

#include <iostream>
#include <vector>
using namespace std;

Move *HumanPlayer::getMove(Game *game) const {
  char input[100];

  cout << "Move for Player " << game->getTurn() << ": ";
  while (true) {
    cin.getline(input, 100);

    string error;
    vector<Move*> moves = game->parseMove(string(input), error);

    //for (Move *m : moves)
    //  cout << m << endl;

    if (moves.size() == 0) {
      if (error.size())
        cout << error << ", try again: ";
      else
        cout << "Invalid move, try again: ";
    }
    else if (moves.size() > 1) {
      cout << "Ambiguous move: ";
    }
    else {
      return moves[0];
    }
  }
}
